docs/architecture/adr-101-data-companion-pull-api.md

```diff
@@ -0,0 +1,403 @@
+# ADR 101: Data Companion Pull API
+
+## Changelog
+
+- 2023-05-03: Update based on synchronous feedback from team (@thanethomson)
+- 2023-04-04: Update based on review feedback (@thanethomson)
+- 2023-02-28: Renumber from 084 to 101 (@thanethomson)
+- 2022-12-18: First draft (@thanethomson)
+
+## Status
+
+Accepted
+
+## Context
+
+See the [context for ADR-100][adr-100-context].
+
+The primary novelty introduced in this ADR is effectively just a new gRPC API
+that allows an external application to influence which data the node prunes.
+Otherwise, existing and planned RPC interfaces (such as the planned gRPC
+interface) in addition to this pruning API should, in theory, deliver the same
+kind of value as the solution proposed in ADR-100.
+
+Even though the pruning API could be useful to operators outside the context of
+the usage of a data companion (e.g. it could provide operators with more control
+over node pruning behaviour), it is presented as part of the data companion
+discussion to illustrate its initial intended use case.
+
+## Alternative Approaches
+
+[ADR-100][adr-100], as well as the [alternatives][adr-100-alt] outlined in
+ADR-100, are all alternative approaches.
+
+## Decision
+
+To implement ADR-101 instead of ADR-100.
+
+## Detailed Design
+
+The model proposed in this ADR inverts that proposed in ADR-100, with the node
+being the server and the data companion being the client. Here, the companion
+"pulls" data from the node.
+
+This provides much weaker data delivery guarantees than the "push" model of
+ADR-100. In this "pull" model, the companion can lag behind consensus, but the
+node does not crash if the companion is unavailable.
+
+### Requirements
+
+The requirements for ADR-101 are the same as the [requirements for
+ADR-100][adr-100-req].
+
+### Entity Relationships
+
+The following model shows the proposed relationships between CometBFT, a
+socket-based ABCI application, and the proposed data companion service.
+
+```mermaid
+flowchart RL
+    comet[CometBFT]
+    companion[Data Companion]
+    app[ABCI App]
+
+    comet --> app
+    companion --> comet
+```
+
+In this diagram, it is evident that CometBFT (as a client) connects out to the
+ABCI application (a server), and the companion (a client) connects to the
+CometBFT node (a server).
+
+### Pruning Behaviour
+
+Two parameters are proposed as a necessary part of the pruning API:
+
+- **Pruning service block retain height**, which influences the height to which
+  the node will retain blocks. This is different to the **application block
+  retain height**, which is set by the application in its response to each ABCI
+  `commit` message.
```

Comment:
```suggestion
  `Commit` message.
```

Reply:
The spec link: https://github.com/cometbft/cometbft/blob/main/spec/abci/abci%2B%2B_methods.md#commit

Reply:
Actually, `[abci-commit]` is a defined link for the document.